server/etcdserver/apply/apply_auth_test.go

```diff
@@ -15,47 +15,844 @@
+	_, err = authApplier.RoleAdd(&pb.AuthRoleAddRequest{Name: roleReadOnly})
+	require.NoError(t, err)
+	_, err = authApplier.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: userReadOnly, Role: roleReadOnly})
+	require.NoError(t, err)
+	_, err = authApplier.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{Name: roleReadOnly, Perm: &authpb.Permission{
+		PermType: authpb.READ,
+		Key:      []byte(key),
+		RangeEnd: []byte(rangeEnd),
+	}})
+	require.NoError(t, err)
+
+	_, err = authApplier.UserAdd(&pb.AuthUserAddRequest{Name: userWriteOnly, Options: &authpb.UserAddOptions{NoPassword: true}})
+	require.NoError(t, err)
+	_, err = authApplier.RoleAdd(&pb.AuthRoleAddRequest{Name: roleWriteOnly})
+	require.NoError(t, err)
+	_, err = authApplier.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: userWriteOnly, Role: roleWriteOnly})
+	require.NoError(t, err)
+	_, err = authApplier.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{Name: roleWriteOnly, Perm: &authpb.Permission{
+		PermType: authpb.WRITE,
+		Key:      []byte(key),
+		RangeEnd: []byte(rangeEnd),
+	}})
+	require.NoError(t, err)
+
+	_, err = authApplier.AuthEnable()
+	require.NoError(t, err)
+}
 
-	b, _ := betesting.NewDefaultTmpBackend(t)
-	defer betesting.Close(t, b)
+// SetAuthInfo manually sets the authInfo of the applier. In reality, authInfo is filled before Apply()
+func SetAuthInfo(authApplier *authApplierV3, userName string) {
+	authApplier.authInfo = auth.AuthInfo{
+		Username: userName,
+		Revision: authApplier.as.Revision(),
+	}
+}
 
-	simpleTokenTTLDefault := 300 * time.Second
-	tokenTypeSimple := "simple"
-	dummyIndexWaiter := func(index uint64) <-chan struct{} {
-		ch := make(chan struct{}, 1)
-		go func() {
-			ch <- struct{}{}
-		}()
-		return ch
+// TestAuthApplierV3_Apply ensures Apply() calls applyFunc() when permission is granted
+// and returns an error when permission is denied
+func TestAuthApplierV3_Apply(t *testing.T) {
+	tcs := []struct {
+		name         string
+		request      *pb.InternalRaftRequest
+		expectResult *Result
+	}{
+		{
+			name: "request does not need admin permission",
+			request: &pb.InternalRaftRequest{
+				Header: &pb.RequestHeader{},
+			},
+			expectResult: &Result{},
+		},
+		{
+			name: "request needs admin permission but permission denied",
+			request: &pb.InternalRaftRequest{
+				Header: &pb.RequestHeader{
+					Username: userReadOnly,
+				},
+				AuthEnable: &pb.AuthEnableRequest{},
+			},
+			expectResult: &Result{
+				Err: auth.ErrPermissionDenied,
+			},
+		},
+		{
+			name: "request needs admin permission and permitted",
+			request: &pb.InternalRaftRequest{
+				Header: &pb.RequestHeader{
+					Username: userRoot,
+				},
+				AuthEnable: &pb.AuthEnableRequest{},
+			},
+			expectResult: &Result{},
+		},
+	}
+
+	authApplier := defaultAuthApplierV3(t)
+	MustCreateRolesAndEnableAuth(t, authApplier)
+	dummyApplyFunc := func(ctx context.Context, r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *Result {
```

Comment:
I think you have this defined twice below, maybe extract this into a variable? like the dummy writer on top